File: best-time-to-buy-and-sell-stock/Solution121.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

class Solution121 {

    // T: O(n), S: O(n)
    public int maxProfit(int[] prices) {
        int n = prices.length;
        var rightMaxes = new int[n];
        rightMaxes[n - 1] = prices[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            rightMaxes[i] = Math.max(rightMaxes[i + 1], prices[i]);
        }
        int maxProfit = 0, currProfit = 0;
        for (int i = 0; i < n; i++) {
            currProfit = rightMaxes[i] - prices[i];
            maxProfit = Math.max(maxProfit, currProfit);
        }
        return maxProfit;
    }

    // T: O(n), S: O(1)
    public int maxProfitOpt(int[] prices) {
        int l = 0, r = 1, n = prices.length;
        int res = 0;
        while (r < n) {
            if (prices[l] < prices[r]) {
                res = Math.max(res, prices[r] - prices[l]);
            } else {
                l = r;
            }
            r++;
        }
        return res;
    }

    public static void main(String[] args) {
        var sol = new Solution121();
        var prices1 = new int[] { 7, 1, 5, 3, 6, 4 };
        var prices2 = new int[] { 7, 6, 4, 3, 1 };
        System.out.println(sol.maxProfit(prices1)); // 5
        System.out.println(sol.maxProfit(prices2)); // 0
        System.out.println(sol.maxProfitOpt(prices1)); // 5
        System.out.println(sol.maxProfitOpt(prices2)); // 0
    }

}

File: best-time-to-buy-and-sell-stock/solution121.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

from typing import List


class Solution121:

    # T: O(n), S: O(n)
    def maxProfit(self, prices):
        n = len(prices)
        right_maxes = [0] * n
        right_maxes[-1] = prices[-1]
        for i in range(n - 2, -1, -1):
            right_maxes[i] = max(right_maxes[i + 1], prices[i])
        max_profit = 0
        for i in range(n):
            curr_profit = right_maxes[i] - prices[i]
            max_profit = max(max_profit, curr_profit)
        return max_profit

    # T: O(n), S: (1)
    def maxProfitOpt(self, prices: List[int]) -> int:
        l, r, n = 0, 1, len(prices)
        res = 0
        while r < n:
            if prices[l] < prices[r]:
                res = max(res, prices[r] - prices[l])
            else:
                l = r

            r += 1

        return res


def main() -> None:
    sol = Solution121()
    prices1 = [7, 1, 5, 3, 6, 4]
    prices2 = [7, 6, 4, 3, 1]
    print(sol.maxProfit(prices1))  # 5
    print(sol.maxProfit(prices2))  # 0
    print(sol.maxProfitOpt(prices1))  # 5
    print(sol.maxProfitOpt(prices2))  # 0


if __name__ == "__main__":
    main()
